File: tests/gas.test.ts
// Copyright (c) Hintents Authors.
// SPDX-License-Identifier: Apache-2.0

import {
    extractGasEstimation,
    budgetToGasEstimation,
    isCpuWarning,
    isCpuCritical,
    isMemoryWarning,
    isMemoryCritical,
    hasBudgetPressure,
    formatGasEstimation,
    BASE_FEE_STROOPS,
    CPU_STROOPS_PER_UNIT,
    MEM_STROOPS_PER_UNIT,
    UPPER_BOUND_MULTIPLIER_PERCENT,
    BudgetUsage,
    SimulationResponse,
    GasEstimation,
} from '../src/gas';

// ─── Test helpers ─────────────────────────────────────────────────────────────

function makeBudget(overrides: Partial<BudgetUsage> = {}): BudgetUsage {
    return {
        cpu_instructions: 50_000_000,
        memory_bytes: 25_000_000,
        operations_count: 10,
        cpu_limit: 100_000_000,
        memory_limit: 50_000_000,
        cpu_usage_percent: 50.0,
        memory_usage_percent: 50.0,
        ...overrides,
    };
}

function makeResponse(overrides: Partial<SimulationResponse> = {}): SimulationResponse {
    return {
        status: 'success',
        budget_usage: makeBudget(),
        ...overrides,
    };
}

// ─── extractGasEstimation ─────────────────────────────────────────────────────

describe('extractGasEstimation', () => {
    it('extracts CPU and memory costs from a valid response', () => {
        const gas = extractGasEstimation(makeResponse());

        expect(gas.cpuCost).toBe(50_000_000);
        expect(gas.memoryCost).toBe(25_000_000);
        expect(gas.cpuLimit).toBe(100_000_000);
        expect(gas.memoryLimit).toBe(50_000_000);
        expect(gas.cpuUsagePercent).toBe(50.0);
        expect(gas.memoryUsagePercent).toBe(50.0);
        expect(gas.operationsCount).toBe(10);
    });

    it('computes fee estimates', () => {
        const gas = extractGasEstimation(makeResponse());

        expect(gas.estimatedFeeLowerBound).toBeGreaterThan(0);
        expect(gas.estimatedFeeUpperBound).toBeGreaterThanOrEqual(gas.estimatedFeeLowerBound);
    });

    it('throws on null response', () => {
        expect(() => extractGasEstimation(null)).toThrow('null or undefined');
    });

    it('throws on undefined response', () => {
        expect(() => extractGasEstimation(undefined)).toThrow('null or undefined');
    });

    it('throws when budget_usage is missing', () => {
        expect(() => extractGasEstimation({ status: 'success' })).toThrow('budget_usage');
    });
});

// ─── budgetToGasEstimation ────────────────────────────────────────────────────

describe('budgetToGasEstimation', () => {
    it('converts BudgetUsage to GasEstimation', () => {
        const bu = makeBudget();
        const gas = budgetToGasEstimation(bu);

        expect(gas.cpuCost).toBe(bu.cpu_instructions);
        expect(gas.memoryCost).toBe(bu.memory_bytes);
        expect(gas.operationsCount).toBe(bu.operations_count);
    });

    it('computes correct lower bound for zero usage', () => {
        const gas = budgetToGasEstimation(
            makeBudget({ cpu_instructions: 0, memory_bytes: 0 }),
        );
        expect(gas.estimatedFeeLowerBound).toBe(BASE_FEE_STROOPS);
    });

    it('computes correct lower bound for CPU-only usage', () => {
        const gas = budgetToGasEstimation(
            makeBudget({ cpu_instructions: 100_000, memory_bytes: 0 }),
        );
        const expectedCpu = Math.floor(100_000 / CPU_STROOPS_PER_UNIT);
        expect(gas.estimatedFeeLowerBound).toBe(BASE_FEE_STROOPS + expectedCpu);
    });

    it('computes correct lower bound for memory-only usage', () => {
        const gas = budgetToGasEstimation(
            makeBudget({ cpu_instructions: 0, memory_bytes: 128 * 1024 }),
        );
        const expectedMem = Math.floor((128 * 1024) / MEM_STROOPS_PER_UNIT);
        expect(gas.estimatedFeeLowerBound).toBe(BASE_FEE_STROOPS + expectedMem);
    });

    it('upper bound = floor(lower * 115 / 100)', () => {
        const gas = budgetToGasEstimation(makeBudget());
        const expected = Math.floor(
            (gas.estimatedFeeLowerBound * UPPER_BOUND_MULTIPLIER_PERCENT) / 100,
        );
        expect(gas.estimatedFeeUpperBound).toBe(expected);
    });
});

// ─── Threshold predicates ─────────────────────────────────────────────────────

describe('threshold helpers', () => {
    const cases: Array<{
        name: string;
        cpuPct: number;
        memPct: number;
        cpuWarn: boolean;
        cpuCrit: boolean;
        memWarn: boolean;
        memCrit: boolean;
        pressure: boolean;
    }> = [
        { name: 'low usage', cpuPct: 30, memPct: 20, cpuWarn: false, cpuCrit: false, memWarn: false, memCrit: false, pressure: false },
        { name: 'CPU warning', cpuPct: 82, memPct: 20, cpuWarn: true, cpuCrit: false, memWarn: false, memCrit: false, pressure: true },
        { name: 'mem warning', cpuPct: 20, memPct: 85, cpuWarn: false, cpuCrit: false, memWarn: true, memCrit: false, pressure: true },
        { name: 'CPU critical', cpuPct: 96, memPct: 20, cpuWarn: true, cpuCrit: true, memWarn: false, memCrit: false, pressure: true },
        { name: 'mem critical', cpuPct: 20, memPct: 97, cpuWarn: false, cpuCrit: false, memWarn: true, memCrit: true, pressure: true },
        { name: 'both critical', cpuPct: 99, memPct: 99, cpuWarn: true, cpuCrit: true, memWarn: true, memCrit: true, pressure: true },
        { name: 'exact warning boundary', cpuPct: 80, memPct: 80, cpuWarn: true, cpuCrit: false, memWarn: true, memCrit: false, pressure: true },
        { name: 'exact critical boundary', cpuPct: 95, memPct: 95, cpuWarn: true, cpuCrit: true, memWarn: true, memCrit: true, pressure: true },
    ];

    it.each(cases)('$name', ({ cpuPct, memPct, cpuWarn, cpuCrit, memWarn, memCrit, pressure }) => {
        const gas: GasEstimation = {
            cpuCost: 0,
            memoryCost: 0,
            cpuLimit: 100_000_000,
            memoryLimit: 50_000_000,
            cpuUsagePercent: cpuPct,
            memoryUsagePercent: memPct,
            operationsCount: 0,
            estimatedFeeLowerBound: 100,
            estimatedFeeUpperBound: 115,
        };

        expect(isCpuWarning(gas)).toBe(cpuWarn);
        expect(isCpuCritical(gas)).toBe(cpuCrit);
        expect(isMemoryWarning(gas)).toBe(memWarn);
        expect(isMemoryCritical(gas)).toBe(memCrit);
        expect(hasBudgetPressure(gas)).toBe(pressure);
    });
});

// ─── formatGasEstimation ──────────────────────────────────────────────────────

describe('formatGasEstimation', () => {
    it('produces a human-readable string', () => {
        const gas = budgetToGasEstimation(makeBudget());
        const s = formatGasEstimation(gas);

        expect(s).toContain('GasEstimation{');
        expect(s).toContain('CPU:');
        expect(s).toContain('Memory:');
        expect(s).toContain('Fee:');
        expect(s).toContain('stroops}');
    });

    it('includes [WARNING] when CPU is elevated', () => {
        const gas = budgetToGasEstimation(
            makeBudget({ cpu_usage_percent: 85, memory_usage_percent: 40 }),
        );
        const s = formatGasEstimation(gas);
        expect(s).toContain('[WARNING]');
    });

    it('includes [CRITICAL] when resources are critical', () => {
        const gas = budgetToGasEstimation(
            makeBudget({ cpu_usage_percent: 96, memory_usage_percent: 97 }),
        );
        const s = formatGasEstimation(gas);
        const count = (s.match(/\[CRITICAL\]/g) || []).length;
        expect(count).toBe(2);
    });
});

// ─── JSON round-trip ──────────────────────────────────────────────────────────

describe('JSON serialization', () => {
    it('round-trips through JSON.stringify / JSON.parse', () => {
        const original = budgetToGasEstimation(makeBudget());
        const json = JSON.stringify(original);
        const parsed: GasEstimation = JSON.parse(json);

        expect(parsed.cpuCost).toBe(original.cpuCost);
        expect(parsed.memoryCost).toBe(original.memoryCost);
        expect(parsed.estimatedFeeLowerBound).toBe(original.estimatedFeeLowerBound);
        expect(parsed.estimatedFeeUpperBound).toBe(original.estimatedFeeUpperBound);
    });
});

File: internal/simulator/gas_test.go
// Copyright 2025 Erst Users
// SPDX-License-Identifier: Apache-2.0

package simulator

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"
)

// ─── ExtractGasEstimation ────────────────────────────────────────────────────

func TestExtractGasEstimation_Success(t *testing.T) {
	resp := &SimulationResponse{
		Status: "success",
		BudgetUsage: &BudgetUsage{
			CPUInstructions:    50_000_000,
			MemoryBytes:        25_000_000,
			OperationsCount:    10,
			CPULimit:           100_000_000,
			MemoryLimit:        50_000_000,
			CPUUsagePercent:    50.0,
			MemoryUsagePercent: 50.0,
		},
	}

	gas, err := ExtractGasEstimation(resp)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if gas.CPUCost != 50_000_000 {
		t.Errorf("CPUCost: want 50000000, got %d", gas.CPUCost)
	}
	if gas.MemoryCost != 25_000_000 {
		t.Errorf("MemoryCost: want 25000000, got %d", gas.MemoryCost)
	}
	if gas.CPULimit != 100_000_000 {
		t.Errorf("CPULimit: want 100000000, got %d", gas.CPULimit)
	}
	if gas.MemoryLimit != 50_000_000 {
		t.Errorf("MemoryLimit: want 50000000, got %d", gas.MemoryLimit)
	}
	if gas.CPUUsagePercent != 50.0 {
		t.Errorf("CPUUsagePercent: want 50.0, got %.2f", gas.CPUUsagePercent)
	}
	if gas.MemoryUsagePercent != 50.0 {
		t.Errorf("MemoryUsagePercent: want 50.0, got %.2f", gas.MemoryUsagePercent)
	}
	if gas.OperationsCount != 10 {
		t.Errorf("OperationsCount: want 10, got %d", gas.OperationsCount)
	}
}

func TestExtractGasEstimation_NilResponse(t *testing.T) {
	_, err := ExtractGasEstimation(nil)
	if err == nil {
		t.Fatal("expected error for nil response")
	}
}

func TestExtractGasEstimation_NoBudget(t *testing.T) {
	resp := &SimulationResponse{Status: "success"}
	_, err := ExtractGasEstimation(resp)
	if err == nil {
		t.Fatal("expected error when BudgetUsage is nil")
	}
}

// ─── Fee estimation arithmetic ───────────────────────────────────────────────

func TestFeeEstimation_Arithmetic(t *testing.T) {
	tests := []struct {
		name         string
		cpuInsns     uint64
		memBytes     uint64
		wantLower    int64
		wantUpperGeq int64 // upper must be >= lower
	}{
		{
			name:      "zero usage",
			cpuInsns:  0,
			memBytes:  0,
			wantLower: BaseFeeStroops, // 100
		},
		{
			name:      "only CPU",
			cpuInsns:  100_000,
			memBytes:  0,
			wantLower: BaseFeeStroops + 10, // 100_000 / 10_000 = 10
		},
		{
			name:      "only memory",
			cpuInsns:  0,
			memBytes:  128 * 1024,
			wantLower: BaseFeeStroops + 2, // 128 KiB / 64 KiB = 2
		},
		{
			name:      "mixed realistic workload",
			cpuInsns:  50_000_000,
			memBytes:  25_000_000,
			wantLower: BaseFeeStroops + 5000 + 381, // 50M/10k + 25M/65536
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			bu := &BudgetUsage{
				CPUInstructions:    tt.cpuInsns,
				MemoryBytes:        tt.memBytes,
				CPULimit:           100_000_000,
				MemoryLimit:        50_000_000,
				CPUUsagePercent:    float64(tt.cpuInsns) / 100_000_000 * 100,
				MemoryUsagePercent: float64(tt.memBytes) / 50_000_000 * 100,
				OperationsCount:    1,
			}

			gas, err := bu.ToGasEstimation()
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}

			if gas.EstimatedFeeLowerBound != tt.wantLower {
				t.Errorf("lower bound: want %d, got %d", tt.wantLower, gas.EstimatedFeeLowerBound)
			}

			if gas.EstimatedFeeUpperBound < gas.EstimatedFeeLowerBound {
				t.Errorf("upper bound (%d) < lower bound (%d)", gas.EstimatedFeeUpperBound, gas.EstimatedFeeLowerBound)
			}

			expectedUpper := gas.EstimatedFeeLowerBound * UpperBoundMultiplierPercent / 100
			if gas.EstimatedFeeUpperBound != expectedUpper {
				t.Errorf("upper bound: want %d, got %d", expectedUpper, gas.EstimatedFeeUpperBound)
			}
		})
	}
}

// ─── Warning / critical thresholds ───────────────────────────────────────────

func TestGasEstimation_ThresholdHelpers(t *testing.T) {
	tests := []struct {
		name               string
		cpuPct             float64
		memPct             float64
		wantCPUWarning     bool
		wantCPUCritical    bool
		wantMemWarning     bool
		wantMemCritical    bool
		wantBudgetPressure bool
	}{
		{"low usage", 30.0, 20.0, false, false, false, false, false},
		{"CPU warning", 82.0, 20.0, true, false, false, false, true},
		{"mem warning", 20.0, 85.0, false, false, true, false, true},
		{"CPU critical", 96.0, 20.0, true, true, false, false, true},
		{"mem critical", 20.0, 97.0, false, false, true, true, true},
		{"both critical", 99.0, 99.0, true, true, true, true, true},
		{"exact warning boundary", 80.0, 80.0, true, false, true, false, true},
		{"exact critical boundary", 95.0, 95.0, true, true, true, true, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gas := &GasEstimation{
				CPUUsagePercent:    tt.cpuPct,
				MemoryUsagePercent: tt.memPct,
			}

			if gas.IsCPUWarning() != tt.wantCPUWarning {
				t.Errorf("IsCPUWarning: want %v, got %v", tt.wantCPUWarning, gas.IsCPUWarning())
			}
			if gas.IsCPUCritical() != tt.wantCPUCritical {
				t.Errorf("IsCPUCritical: want %v, got %v", tt.wantCPUCritical, gas.IsCPUCritical())
			}
			if gas.IsMemoryWarning() != tt.wantMemWarning {
				t.Errorf("IsMemoryWarning: want %v, got %v", tt.wantMemWarning, gas.IsMemoryWarning())
			}
			if gas.IsMemoryCritical() != tt.wantMemCritical {
				t.Errorf("IsMemoryCritical: want %v, got %v", tt.wantMemCritical, gas.IsMemoryCritical())
			}
			if gas.HasBudgetPressure() != tt.wantBudgetPressure {
				t.Errorf("HasBudgetPressure: want %v, got %v", tt.wantBudgetPressure, gas.HasBudgetPressure())
			}
		})
	}
}

// ─── EstimateGas (with MockRunner) ───────────────────────────────────────────

func TestEstimateGas_WithMockRunner(t *testing.T) {
	mock := NewMockRunner(func(req *SimulationRequest) (*SimulationResponse, error) {
		return &SimulationResponse{
			Status: "success",
			BudgetUsage: &BudgetUsage{
				CPUInstructions:    80_000_000,
				MemoryBytes:        30_000_000,
				OperationsCount:    5,
				CPULimit:           100_000_000,
				MemoryLimit:        50_000_000,
				CPUUsagePercent:    80.0,
				MemoryUsagePercent: 60.0,
			},
		}, nil
	})

	req := &SimulationRequest{
		EnvelopeXdr:   "AAAA",
		ResultMetaXdr: "AAAA",
	}

	gas, err := EstimateGas(mock, req)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if gas.CPUCost != 80_000_000 {
		t.Errorf("CPUCost: want 80000000, got %d", gas.CPUCost)
	}
	if gas.MemoryCost != 30_000_000 {
		t.Errorf("MemoryCost: want 30000000, got %d", gas.MemoryCost)
	}
	if !gas.IsCPUWarning() {
		t.Error("expected CPU warning at 80%")
	}
	if gas.EstimatedFeeLowerBound <= 0 {
		t.Errorf("expected positive fee, got %d", gas.EstimatedFeeLowerBound)
	}
}

func TestEstimateGas_NilRunner(t *testing.T) {
	_, err := EstimateGas(nil, &SimulationRequest{})
	if err == nil {
		t.Fatal("expected error for nil runner")
	}
}

func TestEstimateGas_NilRequest(t *testing.T) {
	mock := NewDefaultMockRunner()
	_, err := EstimateGas(mock, nil)
	if err == nil {
		t.Fatal("expected error for nil request")
	}
}

func TestEstimateGas_RunnerError(t *testing.T) {
	mock := NewMockRunner(func(req *SimulationRequest) (*SimulationResponse, error) {
		return nil, fmt.Errorf("binary not found")
	})

	_, err := EstimateGas(mock, &SimulationRequest{EnvelopeXdr: "X", ResultMetaXdr: "Y"})
	if err == nil {
		t.Fatal("expected error when runner fails")
	}
}

func TestEstimateGas_NoBudgetInResponse(t *testing.T) {
	mock := NewDefaultMockRunner() // returns response without BudgetUsage

	_, err := EstimateGas(mock, &SimulationRequest{EnvelopeXdr: "X", ResultMetaXdr: "Y"})
	if err == nil {
		t.Fatal("expected error when budget usage is nil")
	}
}

// ─── String() ────────────────────────────────────────────────────────────────

func TestGasEstimation_String(t *testing.T) {
	gas := &GasEstimation{
		CPUCost:                50_000_000,
		MemoryCost:             25_000_000,
		CPULimit:               100_000_000,
		MemoryLimit:            50_000_000,
		CPUUsagePercent:        50.0,
		MemoryUsagePercent:     50.0,
		OperationsCount:        10,
		EstimatedFeeLowerBound: 5481,
		EstimatedFeeUpperBound: 6303,
	}

	s := gas.String()
	if !strings.Contains(s, "GasEstimation{") {
		t.Errorf("String() missing prefix: %s", s)
	}
	if !strings.Contains(s, "CPU:") {
		t.Errorf("String() missing CPU: %s", s)
	}
	if !strings.Contains(s, "Memory:") {
		t.Errorf("String() missing Memory: %s", s)
	}
	if !strings.Contains(s, "Fee:") {
		t.Errorf("String() missing Fee: %s", s)
	}
}

func TestGasEstimation_String_WithWarning(t *testing.T) {
	gas := &GasEstimation{
		CPUUsagePercent:    85.0,
		MemoryUsagePercent: 40.0,
	}
	s := gas.String()
	if !strings.Contains(s, "[WARNING]") {
		t.Errorf("expected [WARNING] in output: %s", s)
	}
}

func TestGasEstimation_String_WithCritical(t *testing.T) {
	gas := &GasEstimation{
		CPUUsagePercent:    96.0,
		MemoryUsagePercent: 97.0,
	}
	s := gas.String()
	count := strings.Count(s, "[CRITICAL]")
	if count != 2 {
		t.Errorf("expected 2x [CRITICAL], got %d in: %s", count, s)
	}
}

// ─── JSON round-trip ─────────────────────────────────────────────────────────

func TestGasEstimation_JSONRoundTrip(t *testing.T) {
	original := &GasEstimation{
		CPUCost:                42_000_000,
		MemoryCost:             18_000_000,
		CPULimit:               100_000_000,
		MemoryLimit:            50_000_000,
		CPUUsagePercent:        42.0,
		MemoryUsagePercent:     36.0,
		OperationsCount:        7,
		EstimatedFeeLowerBound: 4474,
		EstimatedFeeUpperBound: 5145,
	}

	data, err := json.Marshal(original)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	var decoded GasEstimation
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	if decoded.CPUCost != original.CPUCost {
		t.Errorf("CPUCost mismatch: %d vs %d", decoded.CPUCost, original.CPUCost)
	}
	if decoded.MemoryCost != original.MemoryCost {
		t.Errorf("MemoryCost mismatch: %d vs %d", decoded.MemoryCost, original.MemoryCost)
	}
	if decoded.EstimatedFeeLowerBound != original.EstimatedFeeLowerBound {
		t.Errorf("EstimatedFeeLowerBound mismatch: %d vs %d", decoded.EstimatedFeeLowerBound, original.EstimatedFeeLowerBound)
	}
	if decoded.EstimatedFeeUpperBound != original.EstimatedFeeUpperBound {
		t.Errorf("EstimatedFeeUpperBound mismatch: %d vs %d", decoded.EstimatedFeeUpperBound, original.EstimatedFeeUpperBound)
	}
}

// ─── BudgetUsage.ToGasEstimation ─────────────────────────────────────────────

func TestBudgetUsage_ToGasEstimation(t *testing.T) {
	bu := &BudgetUsage{
		CPUInstructions:    75_000_000,
		MemoryBytes:        40_000_000,
		OperationsCount:    20,
		CPULimit:           100_000_000,
		MemoryLimit:        50_000_000,
		CPUUsagePercent:    75.0,
		MemoryUsagePercent: 80.0,
	}

	gas, err := bu.ToGasEstimation()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if gas.CPUCost != bu.CPUInstructions {
		t.Errorf("CPUCost should equal CPUInstructions")
	}
	if gas.MemoryCost != bu.MemoryBytes {
		t.Errorf("MemoryCost should equal MemoryBytes")
	}
	if gas.OperationsCount != bu.OperationsCount {
		t.Errorf("OperationsCount should match")
	}
	if gas.EstimatedFeeLowerBound <= 0 {
		t.Errorf("fee lower bound should be positive")
	}
	if gas.EstimatedFeeUpperBound < gas.EstimatedFeeLowerBound {
		t.Errorf("fee upper bound should be >= lower bound")
	}
}
